Feature/stage new

jardizzo opened the pull request:
New staging codes with support for multiple YAML data structures describing files to be acquired. Both the "STAGE" file structure and the "GetData" file structure are now supported. Both of these data structures are lists of dictionaries. However, the original "STAGE" file structure is somewhat limiting. The "GetData" file structure allows expanded options as follows:

(1) Support for multiple input source directories
(2) Sentinel parameters for implementing wait logic (minimum file count, minimum file age)
(3) Extensible (easy to add new dictionary parameters and extend the code)
(4) Flexible shortcuts for specifying files or whole directories to be copied
(5) Ability to rename files 

In addition, SWELL exceptions were added as a way of uniquely tagging errors.

Remaining Work:

(1) The current codes still use globs. Need to switch to os.listdir?
(2) GetData remains file-centric (i.e. recursive directory copies are not allowed). The parameter name for specifying items to be copied remains as "files". Entire directories can be specified using wildcards. We need to decide if this needs to be changed (e.g. "items" instead of "files" and allow directory names and recursive copies).

Examples:

(1) Copy all files in directory

src: /data/dir1
dst: /output/dir

(2) Copy random list of files

files:
  - /data/dir1/file1 /output/dirout
  - /data/dir2/file2 /output/dirout2
 
(3) Copy select files or files of select type

src: /data/dir1
dst: /data/output
files:
  - *.yaml
  - *.nc4
  - test.data

(4) Link files

src: /data/dir1
dst: /data/output
link: 1
files:
  - *.yaml

jardizzo committed:
First commit of filehandler.

jardizzo committed:
New version of filehandler and exception modules.

jardizzo committed:
Updated stage task to use new filehandler module.

jardizzo committed:
Changed the factory method name for filehandler.

jardizzo committed:
Need to incorporate the most recent changes to develop for testing.
Merge branch 'develop' into feature/stage_new

jardizzo committed:
Fixed import path for exceptions module.

jardizzo committed:
Activated sentinel logic for the stage task.

danholdaway committed:
Merge branch 'develop' into feature/stage_new

danholdaway commented:
Changes look good @jardizzo and the workflow ran all the way through for me. There's just some coding norms to resolve and then we should merge. You can improve/modify the things you mentioned in some subsequent PRs.

jardizzo committed:
Recommended code-style changes.

jardizzo committed:
Re-syncing with develop.
Merge remote-tracking branch 'refs/remotes/origin/feature/stage_new' into feature/stage_new

jardizzo committed:
Recommended code-style changes.

danholdaway committed:
Merge branch 'develop' into feature/stage_new

jardizzo committed:
Codestyle changes.

jardizzo committed:
Codestyle changes.

jardizzo committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stage_new' into feature/stage_new

jardizzo committed:
Codestyle changes.

jardizzo committed:
Codestyle changes.

jardizzo committed:
Codestyle changes.
